sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/Runner.java

```diff
@@ -1,14 +1,18 @@
 package com.pulumi.deployment.internal;
 
-import com.pulumi.core.Output;
-import com.pulumi.resources.Stack;
-import com.pulumi.resources.StackOptions;
+import com.pulumi.core.internal.annotations.InternalUse;
 
 import java.util.List;
-import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Supplier;
 
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A task {@link CompletableFuture} runner.
+ */
+@InternalUse
 public interface Runner {
```

Comment:
> 1. why does it deal with exit codes at all - at this level?
yes, this is confusing for me as well, I was thinking about moving the error handling out of the runner, or at least out of the mian loop and do it once with a `.handle` not multiple times

> 2. why does the caller need to get swallowed exceptions, [...]
I've removed `getSwallowedExceptions` and have put the exceptions into the `Result`. Right now it is needed for the tests only. Not sure if universally useful. The functionality was not originally there, I've ported it from C# while trying to catch up with upstream changes.

> 4. [...]
I'd love to be able to simplify the signature even more: `CompletableFuture<Void> run()` I'm just not sure if we can guarantee the stack will be created before user code is called just by registering both and making the user's code execution lazy/wait for stack

Then we could make the error handling on the boundary in PulumiInternal, instead of the Runner. 
This way a Runner would do just one thing and be super easy to test and reason about.

I just have a feeling it's not gonna be so easy, but we might try anyway :D

Reply:
I agree, I've gathered the feedback regarding possible future directions in #668 for further considerations.